### ISSUE TITLE
Add Enqueue button in "Always Ask" Share Menu

### ISSUE THREAD

**Issue by MD77MD** [TeamNewPipe/NewPipe#4779]
### Checklist

- [x] I checked, but didn't find any duplicates (open OR closed) of this issue in the repo. 
- [x] I have read the contribution guidelines given at https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md.
- [x] This issue contains only one feature request. I will open one issue for every feature I want to request.

#### Describe the feature you want

Until newPipe 0.20.2, while playing a video in background or pop-up mode, sharing another video from others apps to newPipe would enqueue them directly. However this is changing in the upcoming release because it was determined to be bad UX #4562. therefor I am raising this issue to discuss an  alternative for it.

A direct quote from developer:
>When sharing a video to NewPipe and a play button is pressed, replace the current queue instead of enqueueing. Enqueueing can still be achieved opening info and then enqueueing, but enqueueing directly was bad UX since the user would not see what happened.

#### Is your feature request related to a problem? Please describe it

These are my related conversations with the developer about this issue: https://github.com/TeamNewPipe/NewPipe/pull/4562#issuecomment-712496743, https://github.com/TeamNewPipe/NewPipe/pull/4562#issuecomment-719837705

To summarize: when a video is shared, enqueueing to pop-up or background mode was done directly, however Stypox suggested an alternative to direct enqueueing by shareing video to newPipe, then opening its details from menu, then long pressing enqueue button... however this was a tedious alternative as it would take too much time and effort to do so compared to old method, especially if multiple videos are to be enqueued.. (you can try enqueueing 5 video yourself to see the vast difference).

also when sharing to pop-up or background mode, users actually do not really want video details... they actually just wants the video to be enqueued... if the users indeed wanted video details they would have choosen the first option (video details). ..more so, it results in wasting data for loading the video details page for each video. 

#### Additional context

The alternative for old enqueueing method should be as easy as the old method (enqueueing directly)..  it sould avoid wasting to much time and data.

@opusforlife2 suggested when the share menu appears, long pressing on pop-up or background option would enqueue the video ... my suggestion isn't far away from his but more intuitive if that's the right word. Instead of long pressing, a ( + )"plus sign" or something similar is to be added next to both modes on the right. 
<img alt="Screenshot_20201102-210713" src="https://user-images.githubusercontent.com/63132171/97903007-76578080-1d4f-11eb-999a-3eb4c90819ab.jpg" height=600 />

The work done in #4425  "Replace specific enqueue options with one" could be used to inspire more alternatives. 
#### How will you/everyone benefit from this feature?

Makes life a little bit easier... but seriously, try enqueueing 5 video yourself in both ways to see the vast difference.

**Comment by opusforlife2** [comment]
We don't need anything so complicated. Just an additional "Enqueue" option in the "Always Ask" menu will be enough. It should only show when a stream is open.

Long pressing inside a menu seems like bad UX, anyway.

**Comment by MD77MD** [comment]
i did suggest that too

**Comment by opusforlife2** [comment]
Your title should directly state what you want. Your feature request in one line. Not a description of the situation or what will happen in the future.

**Comment by MD77MD** [comment]
> Your title should directly state what you want. Your feature request in one line. Not a description of the situation or what will happen in the future.

I had hard time describing this. because this problem is not there yet so i had to build from scratch... i think next time I'm just gana pretend it's already there ... building the explanation for the issue was not worth that much time i spent.

**Comment by opusforlife2** [comment]
You can freely explain in your own words in the issue body. The title though has to contain certain keywords so interested users will know they are opening the correct issue.

**Comment by vkay94** [comment]
I think enqueuing is too specific for this dialog. The point of the dialog is to define a default action which should be run automatically while opening a link outside from NewPipe - that's why there is "Just once" and "Always" ;)

Enqueing like this is coupled to the situation that there's already NewPipe playing a stream, therefore this action wouldn't stand "for its own" like the others do (I hope you understand what I mean :'))

**Comment by opusforlife2** [comment]
Good point. What if the "Always" button is greyed out if Enqueue is selected?

**Comment by opusforlife2** [comment]
Or, or, when a user shares a link to Newpipe and another stream is open, we could show a dialogue saying "Stream already playing:

 Play now, Enqueue, Cancel".

**Comment by s1awek** [comment]
Any updates on this?

**Comment by 89z** [comment]
https://github.com/TeamNewPipe/NewPipe/issues/5530

**Comment by 89z** [comment]
This is quite frustrating. The default has been changed to "start a new playlist" instead of "add to playlist". I cant understand how anyone would think that is a good idea. @Stypox can you comment on the justification?

In addition, as shown in the picture, in addition to the default being changed, the interface has no option to change the default back to "add to playlist". I am downgrading until this is addressed.

**Comment by 89z** [comment]
https://github.com/TeamNewPipe/NewPipe/issues/5530

https://github.com/TeamNewPipe/NewPipe/issues/4779
https://github.com/TeamNewPipe/NewPipe/pull/4562

**Comment by 89z** [comment]
So it seems you can do this with newer versions according to @Stypox, but the
steps are awful [1]. First, change your NewPipe default from "Background
player" to "Show info". Then, here is what you have to do **for each video to
be added**:

1. Long press link
2. Click Open With NewPipe App
3. Long press Background

Compared to previous:

1. Long press link
2. Click Open With NewPipe App

I will continue to use the old version [2] until this is resolved.

1. https://github.com/TeamNewPipe/NewPipe/pull/4562#issuecomment-719766107
2. https://github.com/TeamNewPipe/NewPipe/releases/tag/v0.20.2

**Comment by 89z** [comment]
https://github.com/TeamNewPipe/NewPipe/issues/4779
https://github.com/TeamNewPipe/NewPipe/issues/5541
https://github.com/TeamNewPipe/NewPipe/pull/4562

**Comment by MD77MD** [comment]
> So it seems you can do this with newer versions according to @Stypox, but the steps are awful [1]. First, change your NewPipe default from "Background player" to "Show info". Then, here is what you have to do **for each video to be added**:
> 
> 1. Long press link
> 2. Click Open With NewPipe App
> 3. Long press Background
> 
> Compared to previous:
> 
> 1. Long press link
> 2. Click Open With NewPipe App
> 
> I will continue to use the old version [2] until this is resolved.
> 
> 1. [#4562 (comment)](https://github.com/TeamNewPipe/NewPipe/pull/4562#issuecomment-719766107)
> 2. https://github.com/TeamNewPipe/NewPipe/releases/tag/v0.20.2

better yet, you can instal the legacy version 20.2 of newpipe to handle shared links, and normal newpipe (or block adds version) for normal use. this way you can still enjoy the latest version with all its features. 

this is what we have to do when no buddy listens anymore 🙁

**Comment by thinsoldier** [comment]
@MD77MD how do you get 2 different versions of the same app installed at the same time?

**Comment by MD77MD** [comment]
> @MD77MD how do you get 2 different versions of the same app installed at the same time?

actually new publicity is a bit different with a different signature... that's why android consider them aa two different apps because they effectively have two different signatures

this is the link to newpipe legacy version: (or ] can find it yourself from from the TeamNewpipe page;

https://github.com/TeamNewPipe/NewPipe-legacy/releases/tag/v0.20.2

P.S: only up to v0.20.2 still have enqueue, later versions dont lost this feature.

**Comment by SameenAhnaf** [comment]
Never mind. But I think the title should be changed into "Add Enqeue button in "Always Ask" share menu and "Preferred "Open" Action" menu. Sometimes, we are always supposed to enqueue videos the copied link. If no video is being played, it should ask for four ways to be played background, popup, full screen and with video info.

**Comment by atmosfar** [comment]
An unfortunate break in functionality from the earlier versions, hope a simple UX can return. The "official" flow described by @89z is far too convoluted for such an obvious task.

**Comment by s1awek** [comment]
I just don't get why issues about this functionality are being closed one by one without solving it?

**Comment by SameenAhnaf** [comment]
@s1awek Duplicate issues are not required to solve a single problem? A single issue needs to stay open regarding this.

**Comment by s1awek** [comment]
> @s1awek Duplicate issues are not required to solve a single problem? A single issue needs to stay open regarding this.

So what issue is now active? Because so far all the issues I've been watching so far been closed.

**Comment by Bruceforce** [comment]
> > @s1awek Duplicate issues are not required to solve a single problem? A single issue needs to stay open regarding this.
> 
> So what issue is now active? Because so far all the issues I've been watching so far been closed.

This one here you just commented is still active.

**Comment by MD77MD** [comment]
the only thing we're good at here in newpipe is closing issues not solving them lol

**Comment by s1awek** [comment]
@MD77MD Just wondering if Google have some spies here just to destroy development process from inside ;)

**Comment by chouhanyang** [comment]
Hello,

I'm new to this project. However, my daily routine relkes on the "enqueue" to background player heavily, and I actually never use NewPipe UI but sharing video streams from YouTube app instead. My use case looks like:

1. Open YouTube App.
2. Search my favorite talk shows and share to NewPipe background player.
3. Continue enqueue few shows since I'll listen to them while driving.

Since I'm not sure what is the policy of the core UI team. I'll just do a minimum fix since I REALLY need to do this every single morning. :) My fix added "Enqueue background" option to the menu which means "enqueue to background player only" to avoid 
 UI confusion:

![device-2021-03-11-171859](https://user-images.githubusercontent.com/2092412/110890539-15aff580-82a5-11eb-855a-b69460456c2c.png)

You can find the release apk here: https://github.com/SavantOne/NewPipe/releases/tag/v0.20.11-patch

I'll probably patch once a while depending on the office NewPipe release. I cannot rely on old 0.20.2 forever since YouTube may change constantly as we all know. Let me know if you think I should send PR for this change and appreciate if someone from the core team can point me to the test process and/or review process.

**Comment by s1awek** [comment]
You are absolut star! Thank you very much :)

**Comment by MD77MD** [comment]
> Hello,
> 
> I'm new to this project. However, my daily routine relies on the "enqueue" to background player heavily, and I actually never use NewPipe UI but sharing video streams from YouTube app instead. My use case looks like:
> 
> 1. Open YouTube App.
> 2. Search my favorite talk shows and share to NewPipe background player.
> 3. Continue enqueue few shows since I'll listen to them while driving.
> 
> Since I'm not sure what is the policy of the core UI team. I just did a minimum fix since I REALLY needed to do this every single morning. :) My fix added "Enqueue background" option to the menu which means "enqueue to background player only" to avoid
> UI confusion:
> 
> ![device-2021-03-11-171859](https://user-images.githubusercontent.com/2092412/110890539-15aff580-82a5-11eb-855a-b69460456c2c.png)
> 
> You can find the release apk here: https://github.com/SavantOne/NewPipe/releases/tag/v0.20.11-patch
> 
> I'll probably patch once a while depending on the office NewPipe release. I cannot rely on old 0.20.2 forever since YouTube may change constantly as we all know. Let me know if you think I should send PR for this change and appreciate if someone from the core team can point me to the test process and/or review process.

@s1awek this what i was complaining about, how many issues where opened and closed for this simple fix, actually it was a function that was already there and removed by devs. and believe it or not, but i had much talk with dev about not removing it before they did... they told me no one needs it, however as you see people do need it.

anyway, i wish you would not accuse people just like that.

**Comment by s1awek** [comment]
@MD77MD 
> anyway, i wish you would not accuse people just like that.
You referring to "google spies" i mentioned earlier? I did not accuse anyone that was just a joke mate. Maybe a bad one thou.

**Comment by MD77MD** [comment]
> @MD77MD
> 
> > anyway, i wish you would not accuse people just like that.
> 
> You referring to "google spies" i mentioned earlier? I did not accuse anyone that was just a joke mate. Maybe a bad one though.

I see, i guess i took it the wrong way... cheers...

btw thank you @chouhanyang for fix, 

anther workaround i mentioned earlier, is to use  newpipe legacy version 20.2 of newpipe to handle shared links, because best version still has that automatic enqueue function. that is until this fix is merged (god know when😆)

**Comment by thinsoldier** [comment]
> 
> You can find the release apk here: https://github.com/SavantOne/NewPipe/releases/tag/v0.20.11-patch
> 

Installed this 4x now (twice on 2 phones) and I don't see the enqueue option.

**Comment by s1awek** [comment]
@thinsoldier it's probably because now you have 2 NewPipe icons in share menu. Just pick a second one when shearing or you can just completely uninstall original app.

**Comment by SameenAhnaf** [comment]
I suggested a better solution in #5840.

Your wish: Add "Enqeue" option in "Always ask" share memu
My wish: Include a separate option "Auto-enqueue videos shared from other apps if a media is playing" in settings.

Most expectedly, if a media has been already playing, I will enqueue the next video. If a user is never or rarely supposed to enqueue videos, he can disable or enable this option at any time.

There's no need to tripple tap (choose background player or popup again). Rather, double tap (1. Click Share, 2. Choose Newpipe) every time is enough.

There should be no "+" icon beside background player and popup as it may cause accidental touches. Plus, the UI should always stay clean and minimalist.

"Enqeue" option can't be directly included in "Preferred "Open" Action". There will have to be options like "Start playing on the background or enqeue" and "Start playing on popup or enqeue". That actually sounds pretty confusing. A separate option could easily solve the problems.

**Comment by MD77MD** [comment]
> I suggested a better solution in #5840.
> 
> Your wish: Add "Enqeue" option in "Always ask" share memu
> My wish: Include a separate option "Auto-enqueue videos shared from other apps if a media is playing" in settings.
> 
> Most expectedly, if a media has been already playing, I will enqueue the next video. If a user is never or rarely supposed to enqueue videos, he can disable or enable this option at any time.
> 
> There's no need to tripple tap (choose background player or popup again). Rather, double tap (1. Click Share, 2. Choose Newpipe) every time is enough.
> 
> There should be no "+" icon beside background player and popup as it may cause accidental touches. Plus, the UI should always stay clean and minimalist.
> 
> "Enqeue" option can't be directly included in "Preferred "Open" Action". There will have to be options like "Start playing on the background or enqeue" and "Start playing on popup or enqeue". That actually sounds pretty confusing. A separate option could easily solve the problems.

to tell you the truth, both options are needed, as both have there own use case. however i did not dear to ask for both. moreover the feature you are asking was originally there and was oddly removed by one of the dev.

so i would suggest to keep both request, perhaps one of them would go through.

**Comment by SameenAhnaf** [comment]
@MD77MD You are right, bro. Users should be given the flexibility 😄. However, I think "Enqeue Popup" should be added in "Always ask" share menu too.

**Comment by SameenAhnaf** [comment]
@MD77MD You are right, bro. Users should be given the flexibility to do so in both ways. However, "Enqeue background" option should be replaced by only "Enqeue" option. Anyone may ask for enqeuing videos on popup as well.

**Comment by Stypox** [comment]
@chouhanyang thank you for the contribution. Could you open a PR for that? I wasn't able to implement that properly before since there were crashes, could you open a PR so that we can discuss the implementation and merge it?

> this what i was complaining about, how many issues where opened and closed for this simple fix, actually it was a function that was already there and removed by devs. and believe it or not, but i had much talk with dev about not removing it before they did... they told me no one needs it, however as you see people do need it.

@MD77MD Right from the beginning I think everyone agreed this could be added, but since there were problems with the implementation we (I) preferred not to provide a broken implementation, since a workaround is available. @chouhanyang did the right thing: he looked into the problem and found a way to solve it. Your tone instead is just counterproductive: complaining does not help, and one of the reasons I've not contributed much to NewPipe lately is because of such comments. If you think an issue is particularly important but we developers have a different idea, then try to help us anyway you can, e.g. describing the issue perfectly, providing various possible implementations and behaviours, analyzing pros and cons, explaining alternative solutions & showing why your solution is better, taking into consideration all of the consequences of a feature (like a more cluttered ui), ... (ideally learning Java and opening a PR, but that doesn't count). You did some of this, and you helped clear some doubts, but the bad comments just ruin everything. If, when a developer has the time to implement a feature, he finds a well-described one, where an agreement has already been reached between both parties and there is a clear solution that just needs implementing, it is far more likely that he will take it up. Instead, if he finds an issue where the proposed feature could benefit a user group but hurt another, where there is no pros-cons analysis and where there are hateful comments, probably he will just do something else (and maybe even feel frustrated).

An example of what I'm describing is https://github.com/TeamNewPipe/NewPipe/pull/4534#issuecomment-741700417
Before my comment there were many possible implementations available, but it wasn't clear which one would be the best. After my comment, instead:
- three different alternatives have been proposed. Finding clever ideas is something everybody can do, if only they dedicate some time to it. I probably spent one hour thinking (just thinking!) about how to solve #4534 (fun fact: here are the sketches I drew)
- the implementation details for each possible alternative were clear
- pros and cons were analyzed and tradeoffs were clear
- a clear consensous was reached

The only problem was that I had to write that comment myself: if it had been already there I would have probably taken up the issue much earlier and I wouldn't have wasted time on implementing a solution which at the end was discarded. Anyway, at least the discussion was constructive, there were no putting-off comments: if that were the case I would have probably just ignored them and done something else. It is really really frustrating to try to mediate between the needs and the opinions of many people, when nobody is willing to change their mind ever so slightly until the perfect solution

**Comment by chouhanyang** [comment]
@Stypox Thanks for the pointer. A pull request is created and feel free to review/comment. In the mean time, I just setup an dev environment with physical device debugging, and it's pretty fun. Let me know if I could help a bit on other issues if you have in mind. And appreciate your contribution to this project.

**Comment by Stypox** [comment]
Please everyone go vote at https://github.com/TeamNewPipe/NewPipe/pull/5850#issuecomment-802730251

**Comment by Stypox** [comment]
Please everyone go vote at https://github.com/TeamNewPipe/NewPipe/pull/5850#issuecomment-802736043 to choose the best solution :-D

**Comment by mhmdanas** [comment]
It's probably too late for this, but why couldn't the behavior have been kept but with a toast when a video is enqueued? That would solve the issue, right? (Sorry if I missed something or this was suggested before, but I'm not interested in reading all the previous discussion.)

**Comment by MD77MD** [comment]
> It's probably too late for this, but why couldn't the behavior have been kept but with a toast when a video is enqueued? That would solve the issue, right? (Sorry if I missed something or this was suggested before, but I'm not interested in reading all the previous discussion.)

exactly

**Comment by 89z** [comment]
If anyone is still using 0.20.2 [1], I think YouTube finally killed it. I am getting Error reports with every video. Unless a way is available to reset the cache or something to fix it, I think I am screwed, as I dont want to use the newer versions because of this issue. I wrote a YouTube server already, but I might see if I can write a browser add-on as a better solution.

1. https://github.com/TeamNewPipe/NewPipe/releases/tag/v0.20.2

**Comment by MD77MD** [comment]
> It's probably too late for this, but why couldn't the behavior have been kept but with a toast when a video is enqueued? That would solve the issue, right? (Sorry if I missed something or this was suggested before, but I'm not interested in reading all the previous discussion.)

@mhmdanas  is there any hope for this to be  reimplemented?

**Comment by mhmdanas** [comment]
@MD77MD well, #5850 is open, so I would say the answer is an obvious "yes." However, I have no idea _when_ the PR will be merged.

**Comment by 89z** [comment]
As an alternative, if anyone is interested, I have a proof of concept YouTube WebExtension here:

https://github.com/89z/rosso/releases

If you right click on any YouTube link, the extension loads the audio into a new tab. I tested it myself and it does work. Right now it just does one track at a time, but hopefully today or tomorrow I can code up a Enqueue feature.

**Comment by Stypox** [comment]
Here is another simple and conservative proposal. We can just add the "Enqueue" option, but only enable it when the player is open and the queue is ready. If this is not the case, the share menu will be opened even if "Enqueue" is set as the preferred open action. The user will be able to choose there which action to perform.
- This is consistent with the behavior that happens e.g. when "Popup player" is set as the preferred open action, but the user just shared a SoundCloud url for which there is no video: in that case the "Popup player" can't be opened (since there is no video), so the menu will show up.
- This proposal implies a little more time spent on clicking actions for users that always use the same player and set "Enqueue" as the preferred open action. I am talking about users that would like the first shared url to automatically open a specific player and the following ones to automatically enqueue. Those users will have to manually choose the player they wish to open on the first shared url, but from then on the player will be open and so the "Enqueue" action will come into effect, so the urls shared afterwards are all enqueued without having to click actions. Therefore in the end I think it is a good tradeoff.
- This solution is intuitive in my opinion, both because it is consistent with other similar behaviors, and because it is expected that when there is no player open the "Enqueue" can't be performed automatically and therefore further input from the user is needed.

**Comment by Poilaucul** [comment]
<img alt="Mockup" src="https://github.com/TeamNewPipe/NewPipe/assets/2638123/964b06d6-8f56-4d2d-9894-6725898a05cc" width=523 />

**Comment by iAdesanyaDaniel** [comment]
@Stypox 
It really would be nice to have the "enqueue" option in the list 🙂

![Screenshot_20240102-134711](https://github.com/TeamNewPipe/NewPipe/assets/31955090/041a992f-6b67-4788-a05a-9f3e8175b7ca)
![Screenshot_20240102-134719](https://github.com/TeamNewPipe/NewPipe/assets/31955090/4663e1aa-7a52-4a1b-abc5-ee45ec16e109)

**Comment by Anoneom** [comment]
Sorry about the long message. Read the last paragraph first for the most concise version.

Had to type it out to figure out fully what I mean and still think some of the earlier parts of the message could be useful for context. It relates to my understanding of the underlying causes of the misunderstanding between users and devs.

In my opinion the issue here is actually that there should just be one queue from the perspective of the user and it should not be deleted because something else starts playing. If another video is started the queue should be kept to continue after the current video.
This emulates the behaviour of Spotifys and others queue-functions.

If there are three queues they should not be deleted when another is opened, as it deletes input that the user could want to save. I don't see much reason for having separate queues for main and pop-up player though.
There could be users that expect one queue for background and one for viewing, for example for watching video and listening to music/podcasts, but I don't think anyone expects changing between main player and pop-up player to delete the queue. I don't think users appreciate the queue being deleted because of starting to play something in the background player either.

When I send a link I want to either add it to a playlist, play it directly in whatever mode I'm already using or queue it in whatever mode I'm already using. I don't want it to delete my queue because I accidentally picked another player than the one I was using.

The current share menu has many options, but three are probably better encapsulated in the suggested "play button" and the enqueued option is several clicks away where a missclick can delete the current queue. I don't see much reason for the current info button besides queueing something either. I can't imagine the situation where someone sends a link to NewPipe for a reason outside playing now, queueing, adding to playlist or downloading.


I think the share menu would actually be most intuitive with just "play", "enqueue" and "add to playlist" and download, where play opens the content in the main player if nothing is playing and in the current player if something is already playing. For enqueue it will just continue playing in the same mode as it already was once it gets to the queued item
